File: java/src/main/java/vsdk/toolkit/io/context/MaterialContext.java
package vsdk.toolkit.io.context;

public class MaterialContext {
    public int clock; // Incremented each change -- resettable
    public boolean sided; // true if surface is 1-sided, false for 2-sided
    public float nr; // Index of refraction, real and imaginary
    public float ni;
    public float rd; // Diffuse reflectance
    public ColorContext rd_c; // Diffuse reflectance color
    public float td; // Diffuse transmittance
    public ColorContext td_c; // Diffuse transmittance color
    public float ed; // Diffuse emittance
    public ColorContext ed_c; // Diffuse emittance color
    public float rs; // Specular reflectance
    public ColorContext rs_c; // Specular reflectance color
    public float rs_a; // Specular reflectance roughness
    public float ts; // Specular transmittance
    public ColorContext ts_c; // Specular transmittance color
    public float ts_a; // Specular transmittance roughness

    public MaterialContext() {
        clock = 0;
        sided = false;
        nr = 0.0f;
        ni = 0.0f;
        rd = 0.0f;
        rd_c = new ColorContext();
        td = 0.0f;
        td_c = new ColorContext();
        ed = 0.0f;
        ed_c = new ColorContext();
        rs = 0.0f;
        rs_c = new ColorContext();
        rs_a = 0.0f;
        ts = 0.0f;
        ts_c = new ColorContext();
        ts_a = 0.0f;
    }

    public void copy(MaterialContext source) {
        if (source == null) {
            return;
        }
        clock = source.clock;
        sided = source.sided;
        nr = source.nr;
        ni = source.ni;
        rd = source.rd;
        rd_c.copy(source.rd_c);
        td = source.td;
        td_c.copy(source.td_c);
        ed = source.ed;
        ed_c.copy(source.ed_c);
        rs = source.rs;
        rs_c.copy(source.rs_c);
        rs_a = source.rs_a;
        ts = source.ts;
        ts_c.copy(source.ts_c);
        ts_a = source.ts_a;
    }
}

File: typescript/base/src/main/vsdk/toolkit/io/context/MaterialContext.ts
import { ColorContext } from "./ColorContext";

export class MaterialContext {
  public clock: number;
  public sided: boolean;
  public nr: number;
  public ni: number;
  public rd: number;
  public rd_c: ColorContext;
  public td: number;
  public td_c: ColorContext;
  public ed: number;
  public ed_c: ColorContext;
  public rs: number;
  public rs_c: ColorContext;
  public rs_a: number;
  public ts: number;
  public ts_c: ColorContext;
  public ts_a: number;

  public constructor() {
    this.clock = 0;
    this.sided = false;
    this.nr = 0.0;
    this.ni = 0.0;
    this.rd = 0.0;
    this.rd_c = new ColorContext();
    this.td = 0.0;
    this.td_c = new ColorContext();
    this.ed = 0.0;
    this.ed_c = new ColorContext();
    this.rs = 0.0;
    this.rs_c = new ColorContext();
    this.rs_a = 0.0;
    this.ts = 0.0;
    this.ts_c = new ColorContext();
    this.ts_a = 0.0;
  }

  public copy(source: MaterialContext | null): void {
    if (source === null) {
      return;
    }
    this.clock = source.clock;
    this.sided = source.sided;
    this.nr = source.nr;
    this.ni = source.ni;
    this.rd = source.rd;
    this.rd_c.copy(source.rd_c);
    this.td = source.td;
    this.td_c.copy(source.td_c);
    this.ed = source.ed;
    this.ed_c.copy(source.ed_c);
    this.rs = source.rs;
    this.rs_c.copy(source.rs_c);
    this.rs_a = source.rs_a;
    this.ts = source.ts;
    this.ts_c.copy(source.ts_c);
    this.ts_a = source.ts_a;
  }
}
